httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/gss/SpnegoScheme.java

```diff
@@ -0,0 +1,113 @@
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package org.apache.hc.client5.http.impl.auth.gss;
+
+import org.apache.hc.client5.http.AuthenticationStrategy;
+import org.apache.hc.client5.http.DnsResolver;
+import org.apache.hc.client5.http.auth.StandardAuthScheme;
+import org.apache.hc.client5.http.impl.auth.SPNegoScheme;
+import org.ietf.jgss.GSSException;
+import org.ietf.jgss.Oid;
+
+/**
+ * SPNEGO (Simple and Protected GSSAPI Negotiation Mechanism) authentication
+ * scheme.
+ * <p>
+ * This is the new mutual authentication capable Scheme which replaces the old deprecated non mutual
+ * authentication capable {@link SPNegoScheme}
+ * </p>
+ *
+ * <p>
+ * Note that this scheme is not enabled by default. To use it, you need create a custom
+ * {@link AuthenticationStrategy} and a custom
+ * {@link org.apache.hc.client5.http.auth.AuthSchemeFactory}
+ * {@link org.apache.hc.core5.http.config.Registry},
+ * and set them on the HttpClientBuilder.
+ * </p>
+ *
+ * <pre>
+ * {@code
+ * private static class SpnegoAuthenticationStrategy extends DefaultAuthenticationStrategy {
+ *   private static final List<String> SPNEGO_SCHEME_PRIORITY =
+ *       Collections.unmodifiableList(
+ *           Arrays.asList(StandardAuthScheme.SPNEGO
+ *           // Add other Schemes as needed
+ *           );
+ *
+ *   protected final List<String> getSchemePriority() {
+ *     return SPNEGO_SCHEME_PRIORITY;
+ *   }
+ * }
+ *
+ * AuthenticationStrategy spnegoStrategy = new SpnegoAuthenticationStrategy();
+ *
+ * AuthSchemeFactory spnegoFactory = new SpnegoSchemeFactory();
+ * Registry<AuthSchemeFactory> mutualSchemeRegistry = RegistryBuilder.<AuthSchemeFactory>create()
+ *     .register(StandardAuthScheme.SPNEGO, spnegoFactory)
+ *     //register other schemes as needed
+ *     .build();
+ *
+ * CloseableHttpClient mutualClient = HttpClientBuilder.create()
+ *    .setTargetAuthenticationStrategy(spnegoStrategy);
+ *    .setDefaultAuthSchemeRegistry(spnegoSchemeRegistry);
+ *    .build();
+ * }
+ * </pre>
+ *
+ * @since 5.5
+ */
+public class SpnegoScheme extends GssSchemeBase {
+
+    private static final String SPNEGO_OID = "1.3.6.1.5.5.2";
+
+    /**
+     * @since 5.0
+     */
+    public SpnegoScheme(final org.apache.hc.client5.http.auth.gss.GssConfig config, final DnsResolver dnsResolver) {
+        super(config, dnsResolver);
+    }
+
+    public SpnegoScheme() {
+        super();
+    }
+
+    @Override
+    public String getName() {
+        return StandardAuthScheme.SPNEGO;
+    }
+
+    @Override
+    protected byte[] generateToken(final byte[] input, final String gssServiceName, final String gssHostname) throws GSSException {
+        return generateGSSToken(input, new Oid(SPNEGO_OID), gssServiceName, gssHostname);
+    }
+
+    @Override
+    public boolean isConnectionBased() {
+        return true;
```

Comment:
This is wrong and false. If there is only one roundtrip then you don't need to bind the GSS context to the connection otherwise you have to, see also https://github.com/gssapi/mod_auth_gssapi?tab=readme-ov-file#gssapiconnectionbound

Reply:
Right, fixing.
Note that this was copied from the old one.
Should we fix it there as well ?

Reply:
No, leave as is.

Reply:
Sorry, I misread. We should fix the copy, not the original one. My bad!

Reply:
Thinking about this again, This is not a problem when the completion done after one roundtrip. So I guess when the security context says continue needed than this needs to be bound to the connection. WDYT?